src/psfmachine/perturbation.py

```diff
@@ -0,0 +1,478 @@
+        b = np.hstack([0, self.breaks, len(self.time) - 1])
+        for b1, b2 in zip(b[:-1], b[1:]):
+            p = np.arange(b1, b2, self.resolution)
+            if p[-1] != b2:
+                p = np.hstack([p, b2])
+            points.append(p)
+        points = np.unique(np.hstack(points))
+        self.nbins = len(points)
+
+        def func(x, quad=False):
+            """
+            Bins down an input variable to the same time resolution as `self`
+            """
+            if x.shape[0] == len(self.time):
+                return x[points]
+            else:
+                raise ValueError("Wrong size to bin")
+
+        return func
+
+    def _get_bindown_func(self):
+        """Builds a function to lower the resolution of the data through binning"""
+        b = np.hstack([0, self.breaks, len(self.time) - 1])
+        points = np.hstack(
+            [np.arange(b1, b2, self.resolution) for b1, b2 in zip(b[:-1], b[1:])]
+        )
+        points = points[~np.in1d(points, np.hstack([0, len(self.time) - 1]))]
+        points = np.unique(np.hstack([points, self.breaks]))
+        self.nbins = len(points) + 1
+
+        def func(x, quad=False):
+            """
+            Bins down an input variable to the same time resolution as `self`
+            """
+            if x.shape[0] == len(self.time):
+                if not quad:
+                    return np.asarray(
+                        [i.mean(axis=0) for i in np.array_split(x, points)]
+                    )
+                else:
+                    return (
+                        np.asarray(
+                            [
+                                np.sum(i ** 2, axis=0) / (len(i) ** 2)
+                                for i in np.array_split(x, points)
+                            ]
+                        )
+                        ** 0.5
+                    )
+            else:
+                raise ValueError("Wrong size to bin")
+
+        return func
+
+    def pca(self, y, ncomponents=5):
+        """In place operation to add the principal components of `y` to the design matrix
+
+        Parameters
+        ----------
+        y: np.ndarray
+            Input flux array to take PCA of.
+        """
+        if not y.ndim == 2:
+            raise ValueError("Must pass a 2D `y`")
+        if not y.shape[0] == len(self.time):
+            raise ValueError(f"Must pass a `y` with shape ({len(self.time)}, X)")
+
+        # Clean out any time series have significant contribution from one component
+        k = np.ones(y.shape[1], bool)
+        for count in range(3):
+            U, s, V = pca(y[:, k], ncomponents, n_iter=30)
+            k[k] &= (np.abs(V) < 0.5).all(axis=0)
+
+        self._vectors = np.hstack([self._vectors, U])
+        if self.segments:
+            self._vectors = self._cut_segments(self._vectors)
+        self._validate()
+        self._clean_vectors()
+        self.matrix = sparse.csr_matrix(self.bin_func(self.vectors))
```

Comment:
This was a problem because I only tested the `PerturbationMatrix` class with pca components, not the `PerturbationMatrix3d`!! Lesson learned, this now has a fix.